### PR TITLE
"sig-verify-tutorial" for Round 2

### PR DESCRIPTION
**Pull request by aratrika1996**
This pull request is for red·dev's entry "sig-verify-tutorial" for Round 2 of the Avalanche Tutorials contest. This entry was submitted originally under the name of red·dev's CEO, Christopher David Sturgeon ("Kit"). It is a group effort by part of the red·dev team. Today, team member Aratrika Mukherjee is submitting the request. We look forward to your critique and comments! Thank you.

### CODE REVIEW

**Review by cgcardona — CHANGES_REQUESTED** [review]
Why are there 5,000 files changed in this diff? Can you confirm that you have the latest from `master` merged into this PR before I review it? Thanks

**Comment by jpop32** [comment]
Apparently, you included the whole project with the tutorial. There's no need for that, you can only include the source files. My guess is (not a js dev myself) that for example the whole node_modules directory can be omitted and recreated locally when someone pulls the project.

**Comment by aratrika1996** [comment]
I have made the required changes and removed all the unnecessary files including the node modules. Kindly review the PR. If you need to review any of the files that I removed, the original repo is here: https://github.com/red-dev-inc/sig-verify-tutorial

**Comment by kitpub** [comment]
Hello, we're ready to make any requested changes. Aratrika is working on IST and I'm on EDT just fyi. We overlap each morning between 8 AM and 10:30 AM EDT.

**Comment by aratrika1996** [comment]
Renamed png for standards compliance per jpop32's comment on our other tutorial's pull request.

**Comment by kitpub** [comment]
Hello! Are revisions due today? If so, we think we've made the ones needed, but please confirm. If you would like to request more changes, we would be happy to make additional edits, but we will need a day or two more to complete them due to time zone differences, etc. Thanks. We've really enjoyed creating material for this contest.

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
+# Tutorial: Avalanche Signature Verification in a dApp
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:1]: * [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md022)
* [Multiple top-level headings in the same document](https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md025)

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+(**Note:** If you're already a Solidity coder, you might think that there is an easier way to do this using the EVM's built-in function **ecrecover**. However, there is one small hitch that makes using **ecrecover** impossible: it uses a different hashing method. While Avalanche uses SHA-256 followed by ripemd160, the EVM uses Keccak-256.)
+
+We have set up a [demo webpage here](https://rediyeti.com/avax-sig-verify-demo).
+
+## Requirements
+
+Metamask needs to be installed on your browser, and you need to be connected to the Avalanche Fuji test network (for this tutorial). You can add a few lines of codes to check if your browser has Metamask installed, and if installed, then to which network you are connected. For instance:
+
+```javascript
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:36]: Please port from javascript -> typescript

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+3. The x and y coordinates of the public key and the 33-byte compressed public key
+
+We extract them using Javascript instead of in the dApp because the Solidity EC library needs them to be separated, and it is easier to do it in Javascript. There is no security risk in doing it here off-chain as we can verify in the dApp that they are indeed related to each other, returning a signature failure if they are not.
+
+### 1. Hash the message
+
+First, you will need to hash the original message. Here is the standard way of hashing the message based on the Bitcoin Script format and Ethereum format:
+
+
+```
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:73]: What is the data type of this code?

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+sha256(length(prefix) + prefix + length(message) + message)
+```
+
+The prefix is a so-called "magic prefix" string `\x1AAvalanche Signed Message:\n`, where `0x1A` is the length of the prefix text and `length(message)` is an integer of the message size. After concatenating these together, hash the result with `sha256`. For example:
+
+```javascript
+function hashMessage() {
+    let mBuf = Buffer.from(message, 'utf8')
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:81]: Add type to each variable declaration after porting to typescript.

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    return result;
+}
+```
+Let's take a closer look at its supporting functions and key features.
+### Bech32 Encoding
+
+We have ported Bech32 to Solidity from the [Bech32 Javascript library](https://github.com/bitcoinjs/bech32).
+There are four functions, **polymod**, **prefixChk**, **encode** and **convert**, used to convert to Bech32 address. 
+
+```
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:235]: Data type of code block?

File: build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+### Verify X-Chain address
+
+It is a simple step, but it is very important to check to see if the extracted X-Chain address from the public key matches with the X-Chain address that was passed from the webapp. Otherwise, you may have a perfectly valid message signature but for a _different_ X-Chain address than the webapp requested. Only if they match can you proceed to verify the signature. Otherwise, return an error message.
+
+### Validate signature
+For verifying the signature, we start with this [Solidity project on Elliptic Curve](https://github.com/tdrerup/elliptic-curve-solidity). 
+
+However, this project uses the **secp256r1** curve. As Avalanche uses **secp256k1** curve, we need to modify the constant values based on this curve. (**Note:** Since modifying cryptographic functions is risky, these are the only modifications we have made.) The constants now look like this:
+
+```
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-sig-verify-tutorial/README.md:314]: Data type of this block of code?

**Review by cgcardona — CHANGES_REQUESTED** [review]
Great work. Minor changes requested.

**Comment by aratrika1996** [comment]
I have made all the required changes to the best of my understanding. Please review the PR and kindly let us know if any more revisions are required.

**Comment by kitpub** [comment]
We are so honored to have won and to have had this chance to contribute to Avalanche! Thank you so much for the opportunity. The team at red·dev is looking forward to much more to come as the ecosystem grows, and we are 100% committed to its success. Please let us know how we can help going forward.

**Comment by kitpub** [comment]
One follow-up question. Regarding the signature verification process we describe in this tutorial, we believe it is correct. However, we are not cryptographers. We are just software devs who have some experience with applied cryptography and love building on Avalanche. I'm wondering what level of auditing for correctness our tutorial received? Since a defect in it could allow an attacker to drain large amounts of AVAX from our dapp, it's a concern, especially now since others might follow this tutorial and use the same technique for their dapps.

On a related note, we also did a pull request for the ECDSA repo we used (and modified slightly). https://github.com/tdrerup/elliptic-curve-solidity/pull/3 As you can see, they have not had time to look at it yet and appear to be busy with other things.

**Pull request merged**